Add missing formatting sections (#396)
unicode-org/message-format-wg: spec/formatting.md

> # DRAFT MessageFormat 2.0 Formatting Behaviour

Author: Ach, that's hard as the speller in question is between my ears. But dropped "behaviour" nevertheless.

## Introduction

This document defines the behaviour of a MessageFormat 2.0 implementation
when formatting a message for display in a user interface, or for some later processing.

> Message formatting is defined in terms of the following stages:

Reviewer: I'm not sure if "stages" is the right way to think of this. In particular, some "stages" contain (or can be thought of as-if they contain) another stage. Perhaps "operations" would be more useful here?

Author: Switched to "operations" for now. Not sure it's really right, but it's not horrible either.

> - **_Resolution_** determines the value of a part of the message,

Reviewer: You might be missing an "assignment" stage (processing of `let` statements), which occurs before selection (and might contain some "resolutions")?

Author: That's done implicitly through variable resolution. There's also this on line 736:

> Resolution and Formatting errors in _expressions_ that are not used
> in _pattern selection_ or _formatting_ MAY be ignored
> as such do not impact the current message's formatting.

In other words, an "assignment" stage is explicitly left out to allow for implementations that either:
1. Eagerly resolve all declarations immediately, or
2. Lazily only resolve declarations that are required by the message.

If we did include an assignment stage, we would effectively mandate the former even when an implementation could determine that it never needed the value of a declaration.

  such as an _expression_, with reference to the current _formatting context_.

>   This may include multiple steps,

Reviewer: In general, avoid normative keywords unless used normatively

```suggestion
This can include multiple steps, 
```

  such as looking up the value of a variable and calling formatting functions.
  The resolved value is not necessarily in the shape it will finally take,
  but is "formattable", i.e. it contains everything required by the eventual formatting.

  > For example, the resolved value of the _expression_ `{|0.40| :number style=percent}`
  > could be an object such as
  >
  > ```
  > { value: Number('0.40'),
  >   formatter: NumberFormat(locale, { style: 'percent' }) }
  > ```
  >

>   > The shape of this value is completely implementation-dependent,

Reviewer: "completely" is unnecessary. Remove.

>   > and different implementations may choose to perform different levels of resolution.

Reviewer: here `MAY` is normative and should be formatted normatively

Author: Hmmm. This should probably be reformulated a bit differently, as it's currently in an example's `>` block which ought to mean it's non-normative.

- **_Pattern Selection_** determines which of a message's _patterns_ is formatted.
  For a message with no _selectors_, this is simple as there is only one _pattern_.

>   With _selectors_, this will depend on their _resolution_.

Reviewer: instead of selectors doing resolution, why not point to the section about selection?

Author: The intent here is to establish a dependency of pattern selection on resolution.

- **_Formatting_** takes the resolved values of the selected _pattern_,
  and formats them in the desired shape.
  This specification only defines formatting messages as a single concatenated string,
  but implementations SHOULD provide formatters for additional shapes

>   as appropriate for their setting.

Reviewer: This is what I was mentioning in the call. We should start to think in terms of `formatToParts`. We can't say how the implementation works or what the name of objects/classes/data structures are. But we can describe formatting as resolving the _message_ into a logical sequence of (um, erm) values. 

---

A formatted message as a whole has some properties associated with it: locale, base direction, and a sequence of "parts"

Each "part" also has a set of properties. The parts are in a logically ordered sequence or array. Each part has a locale and a base direction property. Additional properties MAY be defined by the implementation.

There are two kinds of "part": a "literal part" and an "expression" part.

Each "literal part" consists of a string.

An "expression part" can be resolved to a sequence of zero or more "literal parts".

The string output of a message is the concatenated sequence of resolved literal parts. 

> Here is a simple terrible example:
> Inputs:
> | What | Value | Description |
> |---|---|---|
> | Locale | ar-AE | Locale to use for formatting |
> | `date` | `2023-06-19` | Data value passed to formatter |
>
> Message:
> 
> `{The example date is {$date :datetime skeleton=yMMMd}}`
>
> Output:
> ```
> {
>    "locale": "ar-AE",
>    "direction": "ltr",
>    "parts": [
>        {
>            "type": "literal",
>            "locale": "ar-AE",
>            "direction": "ltr",
>            "value": "The example date is "
>        },
>        {
>            "type": "expression",
>            "locale": "ar-AE",
>            "direction": "rtl",
>            "value": [
>                   { "type": "literal", "locale":"ar-AE", "dir": "rtl", "name": "day", "19" },
>                   { "type": "literal", "locale":"ar-AE", "dir": "rtl", "name": "separator", " " },
>                   { "type": "literal", "locale":"ar-AE", "dir": "rtl", "name": "month", "يونيو" },
>                   { "type": "literal", "locale":"ar-AE", "dir": "rtl", "name": "separator", " " },
>                   { "type": "literal", "locale":"ar-AE", "dir": "rtl", "name": "year", "2023" },
>            ]
>        },
>    ]
> }
> ```

Callers can consume the sequence of parts in order to perform additional processing, such as markup. The above example might be formatted into an HTML context thusly:

> ```html
> <p lang="ar-AE" dir="ltr">The example date is 
>     <span dir="rtl" id="date">19 <em>يونيو</em> 2023</span></p>
> ```

Author: > Each "literal part" consists of a string.
> 
> An "expression part" can be resolved to a sequence of zero or more "literal parts".

This is not always true. For one example, consider this message:
```
{This is my image: {$img}. Isn't it pretty?}
```

When formatting this, the value of `$img` could be a representation of the image itself, rather than any sequence of literal parts.

Next, consider this message:
```
{This is my image: {flower.png :image}. Isn't it pretty?}
```

Here, the message doesn't include any variables, but it does make use of a custom `:image` function, which could format as a representation of the image that is similarly non-stringifiable.

In other words, we cannot make any assumptions about the shape of external variables or the return values of custom functions.

Reviewer: Message format can have parts that are objects.

Literal parts of a message are *always* strings by definition.

Expressions can, as you note, represent objects and these might not be immediately stringable. But the existence of string resolution (note your text does this!!) means that all expressions can ultimately be represented as a sequence of literals. It is tempting to want to make it a list of literals *or* expressions. But ultimately what my text says is that you **can** resolve the deepest nesting of an expression to a string.

The string might be `<img src="flower.png">` or `img/png:base64gooHere` or something. And users do not have to force the object to become a string (they can peek at the expression "type" and get the image, for example).

I'm open to a lot of change here: my proposal above is basically the back of my cocktail napkin while thinking about message resolution. Allowing "shapes" other than string is fine, but having our standard require that one be able to produce a character sequence means that everything can ultimately call `toString`.

To @macchiati's point, "shape" is kind of a vague word. Perhaps:

```suggestion
- **_Formatting_** takes the resolved values of the selected _pattern_,
  and returns the formatted result for the _message_.
  This specification defines formatting of each _message_ as a _string_.
 Implementations MAY return a _message_ using a different, locally appropriate, 
 data type (such as an attributed string) or as a logical sequence of 
 values as appropriate for that implementation.
 
 > For example, an implementation might choose to return an interstitial
 > object so that the caller can "decorate" portions of the formatted value.
 > The `NumberFormatter` class in ICU4J, for example, returns a `FormattedNumber`
 > object, so a _pattern_ such as `{This is my number {42 :number}}` might return
 > the character sequence `This is my number ` followed by a `FormattedNumber`
 > object representing the value `42` in the current locale.
```

Author: I've taken some of the suggested language and incorporated it into the Formatting section, rather than this Introduction. Also moved out & refactored the resolution example slightly.

their handling during formatting is specified here as well.

Formatting of a _message_ is defined by the following operations:

- **_Resolution_** determines the value of a part of the message,
  such as an _expression_, with reference to the current _formatting context_.
  This can include multiple steps,
  such as looking up the value of a variable and calling formatting functions.
  The resolved value is not necessarily in the shape it will finally take,
  but is "formattable", i.e. it contains everything required by the eventual formatting.

  > For example, the resolved value of the _expression_ `{|0.40| :number style=percent}`
  > could be an object such as
  >
  > ```
  > { value: Number('0.40'),
  >   formatter: NumberFormat(locale, { style: 'percent' }) }
  > ```
  >
  > The shape of this value is implementation-dependent,
  > and different implementations may choose to perform different levels of resolution.

- **_Pattern Selection_** determines which of a message's _patterns_ is formatted.
  For a message with no _selectors_, this is simple as there is only one _pattern_.
  With _selectors_, this will depend on their _resolution_.

> - **_Formatting_** takes the resolved values of the selected _pattern_,

Reviewer 2: The earlier definition of "resolution" said "**_Resolution_** determines the value of a part of the message", but this definition of "formatting" implies that the output of resolution is one or more values. So for consistency, either the earlier definition should say "values" instead of "value", or this definition should rely on a shared definition of "value" that includes multiple parts.

Author: The plural here is intended to refer to the distinct _text_ and _expression_ parts of a _pattern_. Each such part of _pattern_ would still resolve to one value each.

Reviewer 2: In that case, maybe it's worth adding something after line 15 saying that the result of resolving a pattern is a list of values that results from independently resolving each of its parts? If the definition of formatting a pattern requires a list as input, then maybe it's worth saying that the output of resolving a pattern is a list, even if the shape of a "value" is being left abstract.

Reviewer 2: In that case, maybe it's worth adding something after line 15 saying that the result of resolving a pattern is a list of values that results from independently resolving each of its parts? If the definition of formatting a pattern requires a list as input, then maybe it's worth saying that the output of resolving a pattern is a list, even if the shape of a "value" is being left abstract.

Author: Your comments lead me to realise that really practically all of the "resolution" section is about "expression resolution". So I'm retitling accordingly, hopefully adding some clarity here as well.

and the observable behavior of the formatter matches that described here.

## Formatting Context

A message's **_formatting context_** represents the data and procedures that are required
for the message's _expression resolution_, _pattern selection_ and _formatting_.

At a minimum, it includes:

- Information on the current **_locale_**,
  potentially including a fallback chain of locales.
  This will be passed on to formatting functions.

- Information on the base directionality of the message and its _text_ tokens.

>   This will be used by strategies for bidirectional isolation.

Reviewer: Not just! It should also be exposed as a property of the message so that the caller can properly set the base paragraph direction. Or is this what you mean by "strategies for bidirectional isolation"? If the latter, I'd suggest:

```suggestion
This can be used by callers to set the base direction of the _message_ upon display.
```

Author: Included in #411.

A message's **_formatting context_** represents the data and procedures that are required
for the message's _resolution_, _pattern selection_ and _formatting_.

At a minimum, it includes:

- Information on the current **_locale_**,
  potentially including a fallback chain of locales.
  This will be passed on to formatting functions.

- Information on the base directionality of the message and its _text_ tokens.
  This will be used by strategies for bidirectional isolation.

- A mapping of string identifiers to values,

>   defining variable values that may be used during _variable resolution_.

Reviewer 2: Is it worth saying here that this mapping is for "external variables", as distinct from the variables defined with `let`-declarations?

Author: I'd prefer not adding to that here atm, as variable resolution is still being iterated upon.

implementations SHOULD NOT immediately fully format the value for output.

In _selectors_, the resolved value of an _expression_ is used for _pattern selection_.

In a _pattern_, the resolved value of an _expression_ is used in its _formatting_.

Depending on the presence or absence of an _operand_
and a _function_ or _reserved_ _annotation_,
one of the following is used to resolve the value of the _expression_:

- If the _expression_ contains no _annotation_,
  its resolved value is determined by _literal resolution_ or _variable resolution_,
  depending on the shape of the _operand_.
- Else, if the _expression_ has a _function_ _annotation_,
  its resolved value is defined by _function resolution_.
- Else, the _expression_ has a _reserved_ _annotation_,
  and a fallback value is used as its value.

### Literal Resolution

The resolved value of a _text_ or a _literal_ is
the character sequence of the _text_ or _literal_
after any character escape has been converted to the escaped character.

When a _literal_ is used as an _operand_
or on the right-hand side of an _option_,
the formatting function MUST treat its resolved value the same

> independently of whether the value was originally _quoted_ or _unquoted_.

Reviewer 2: ```suggestion
whether its value was originally _quoted_ or _unquoted_.
```

> For example,
> the _option_ `foo=42` and the _option_ `foo=|42|` are treated as identical.

> The resolution of a _text_ or _literal_ token MUST always succeed.

Reviewer 2: I don't think the term "token" has been defined yet.

Author: It's from here: https://github.com/unicode-org/message-format-wg/blob/main/spec/syntax.md#tokens

Reviewer 2: Can that be cited in this document?

Author: I believe it effectively already is, as we've established in `spec/README.md` that e.g. the italic "_text_" is a reference to the bold-italic "**_text_**" definition that's a subsection of the "Tokens" section of `syntax.md`.

I don't know how exactly we'll do the final rendering of this, but I would presume that we would at that stage linkify these terms accordingly.

It is an error for a local variable definition to
refer to a local variable that's defined after it in the message.

The resolution of a _variable_ MAY fail if no value is identified for its _name_.
If this happens, an Unresolved Variable error MUST be emitted.
If a _variable_ would resolve to a _fallback value_,
this MUST also be considered a failure.

### Function Resolution

> To resolve the value of an _expression_ with a _function_ _annotation_,

Reviewer 2: ```suggestion
To resolve an _expression_ with a _function_ _annotation_,
```
As I understand the earlier definition, values aren't resolved; expressions are resolved to values.

the following steps are taken:

1. If the _expression_ includes an _operand_, resolve its value.
   If this fails, use a _fallback value_ for the _expression_.
2. Based on the _function_ starting sigil and _name_,
   find the appropriate function implementation from the _function registry_.

>    If this fails, emit an Unknown Function error

Reviewer 2: ```suggestion
   If the registry does not define an implementation for this name, emit an Unknown Function error
```

   and use a _fallback value_ for the _expression_.

> 3. Resolve the _option_ values as a mapping of string identifiers to values.

Reviewer 2: ```suggestion
3. Resolve the _option_ values to a mapping of string identifiers to values.
```

>    For each _option_, resolve its right-hand side value.
>    If this succeeds, assign it to the mapping with the _name_ of the _option_.
>    If this fails, do not set any value in the resolved mapping for this _option_.

Reviewer 2: ```suggestion
   For each _option_:
     * If its right-hand side successfully resolves to a value, bind the _name_ of the _option_ to the resolved value in the mapping.
     * Otherwise, do not bind the _name_ of the _option_ to any value in the mapping.
```

4. Call the function implementation with the following arguments:

   - The current _locale_.
   - The resolved value of the _operand_, or if there is no _operand_,

>      an empty value.

Reviewer 2: Not sure what "an empty value" means concretely.

Author: It would be something like `undefined`, `null`, or `void`, depending on the programming language.

Reviewer 2: There are programming languages that don't have any of those, so perhaps change to something like "a distinct argument that denotes the absence of a value, which depends on the underlying implementation and programming language"?

Author: Fair point. Rephrasing to avoid the term and make the operand an optional argument.

   - The resolved mapping of _options_.

>    The shapes of the resolved _operand_ and _option_ values is implementation-defined.

Reviewer 2: ```suggestion
   The shapes of the resolved _operand_ and _option_ values are implementation-defined.
```

>    An implementation MAY include additional arguments for the function call,

Reviewer 2: ```suggestion
   An implementation MAY pass additional arguments to the function,
```

   as long as reasonable precautions are taken to keep the function interface
   simple and minimal, and avoid introducing potential security vulnerabilities.

   As implementations MAY allow custom functions to be defined by users,
   their access to the _formatting context_ SHOULD be minimal and read-only,
   and their execution time SHOULD be limited.

5. If the call succeeds,
   resolve the value of the _expression_ as the result of that function call.
   If the call fails or does not return a valid value,
   emit a Resolution error and use a _fallback value_ for the _expression_.

### Fallback Resolution

The _resolution_ of an _expression_ may fail in the following cases:

> - A failure in the resolution of a _variable_ _operand_.
> - A failure in the resolution of a _function_ _annotation_.
> - The presence of a _reserved_ _annotation_.

Reviewer 2: ```suggestion
- A _variable_ _operand_ fails to resolve.
- A _function_ _annotation_ fails to resolve.
- The _expression_ has a _reserved_ _annotation_.
```

> In each such case, and error MUST be emitted

Reviewer 2: ```suggestion
In each such case, an error MUST be emitted
```

and a **_fallback value_** used for the _expression_.
This value depends on the shape of the _expression_:

- _expression_ with _literal_ _operand_: U+007C VERTICAL LINE `|`
  followed by the value of the Literal,
  and then by U+007C VERTICAL LINE `|`.
  The same representation is used for both _quoted_ and _unquoted_ values.

  > Examples: `|your horse|`, `|42|`

- _expression_ with _variable_ _operand_: U+0024 DOLLAR SIGN `$`
  followed by the _variable_ _name_ of the _operand_

  > Example: `$user`

- _expression_ with no _operand_:
  the _function_ starting sigil followed by its _name_

  > Examples: `:platform`, `+tag`, `-tag`

- Otherwise: The U+FFFD REPLACEMENT CHARACTER `�` character

_Option_ names and values are not included in the _fallback value_.

When an error occurs in an _expression_ with a _variable_ _operand_
and the _variable_ refers to a local _declaration_,

> the _fallback value_ is formatted based on the _expression_ of the _declaration_,

Reviewer 2: ```suggestion
the _fallback value_ is formatted based on the _expression_ on the right-hand side of the _declaration_,
```

rather than the _expression_ in the _selector_ or _pattern_.

> For example,

> > attempting to format either of the following messages within a context that
> > does not provide for the function `:func` to be successfully resolved:

Reviewer 2: ```suggestion
> in a context in which the function `:func` fails to resolve,
> attempting to format either of the following messages:
```

>
> ```
> let $var = {|horse| :func}
> {The value is {$var}.}
> ```
>
> ```
> let $var = {|horse|}
> {The value is {$var :func}.}
> ```
>

> > would in both cases result in the _pattern_ _expression_
> > resolving to a _fallback value_ of `|horse|`.

Reviewer 2: ```suggestion
> yields a _fallback value_ of `|horse|`.
```

Author: The current phrasing is intended to refer only to the result of the resolution, rather than commenting on the result of the formatting. With "yields", this distinction is lost.

> _Pattern selection_ is not supported for _fallback values_.

Reviewer 2: ```suggestion
_Pattern selection_ is not defined on _fallback values_.
```

Author: The current wording matched that in "Resolve Selectors", where step 2.ii. asks "If selection is supported for `rv`". If you think this should be changed, then that method line should also be changed correspondingly.

Reviewer 2: I'll withdraw this comment. As long as it's consistent, "supported for" is fine.

## Pattern Selection

> When a _message_ contains a _match_ statement with one or more _expressions_,

Reviewer 2: ```suggestion
When a _message_ contains a _match_ construct with one or more _expressions_,
```

4. The pattern `{Exact match}` of the most preferred `1` variant is selected.

## Formatting

After _pattern selection_,

> each _text_ and _expression_ of the selected _pattern_ must be resolved and formatted.

Reviewer 2: ```suggestion
each _text_ and _expression_ part of the selected _pattern_ must be resolved and formatted.
```

(Not sure if we have a standard name for the parts/elements/components of a pattern yet.)

Author: We don't; that's another possible addition to the syntax like `operand` that we might want to make for the sake of clarity.

_Formatting_ is a mostly implementation-defined process,
as it depends on the implementation's shape for resolved values
and the target shape of the formatting.

Formatting errors MAY be emitted during _formatting_,

> as the successful _resolution_ of an _expression_ does not guarantee success in its _formatting_.

Reviewer 2: ```suggestion
as formatting is not defined on every resolved value.
```

A formatter MAY provide a value to use in such a case instead of a _fallback value_.

_Formatting_ MAY produce formatted messages in the following shapes,
as well as any others:

- A single concatenated string.
- A flat sequence of objects corresponding to each resolved value.
- A hierarchical structure of objects that group spans of resolved values,
  such as sequences delimited by "open" and "close" _function_ _annotations_.

Implementations SHOULD provide _formatting_ output shapes that match user needs,

> such that e.g. re-parsing a message formatted as a string is not required.

Reviewer 2: Not sure I understand this -- should it be "re-parsing an output shape that is a single formatted string"?

Author: I think so? As in, there should be no need for someone to format MF2 to a string, and then right away try to parse that string.

Reviewer 2: I understand now, but I still find the text as it is a bit hard to follow. Maybe it would be easier to read split into two sentences, like:

```
Implementations SHOULD provide _formatting_ output shapes that match user needs.
For example, a user should never need to immediately re-parse the string output by the formatter.
Instead, the formatter itself should directly produce a more structured representation.
```

That's still a bit verbose, but the point to get across is that `string` shouldn't serve as an intermediate data structure, I think.

Author: Would this be clearer?
```suggestion
including situations that require further processing of formatted messages.
```

Author: Applied the above change.

    > when * 1 {Second is one}
    > ```

>   - **Duplicate Option Name errors** occur when the same _name_ is used
>     for more than one _option_ in the same _expression_.

Reviewer 2: ```suggestion
  - **Duplicate Option Name errors** occur when the same _name_ 
    appears on the left-hand side
    of more than one _option_ in the same _expression_.
```

  > {Your {$field} is {$id :get field=$field}}
  > ```

Syntax and Data Model errors MUST be emitted as soon as possible.

During selection, an _expression_ handler MUST only emit Resolution and Selection errors.
During formatting, an _expression_ handler MUST only emit Resolution and Formatting errors.

Resolution and Formatting errors in _expressions_ that are not used

> in _pattern selection_ or _formatting_ MAY be ignored
> as such do not impact the current message's formatting.

Reviewer 2: ```suggestion
in _pattern selection_ or _formatting_ MAY be ignored,
as they do not affect the output of the formatter.
```

In all cases, when encountering an error,
a message formatter MUST provide some representation of the message.
An informative error or errors MUST also be separately provided.

When a message contains more than one error,
or contains some error which leads to further errors,
an implementation which does not emit all of the errors
SHOULD prioritise Syntax and Data Model errors over others.

When an error occurs in the _resolution_ of an _option_,
the surrounding _expression_ MUST be processed as if the _option_ were not present.

> This MAY allow for the _expression_ _resolution_ not resulting in a _fallback value_,

Reviewer 2: ```suggestion
This MAY allow the _expression_ to resolve to a non-fallback _value_,
```

though an error MUST still be emitted.

When an error occurs within a _selector_,
the _selector_ MUST NOT match any _variant_ _key_ other than the catch-all `*`

> and a Resolution or Selector error is emitted.

Reviewer 2: Should this be "MUST BE emitted"?

Author: Yes.
```suggestion
and a Resolution or Selector error MUST be emitted.
```
